src/main/java/com/puppycrawl/tools/checkstyle/filters/PatchXpathFilterElement.java

```diff
@@ -183,10 +196,14 @@ private boolean lineMatching(int currentLine) {
     private boolean isMatchingByContextStrategy(TreeWalkerAuditEvent event) {
         boolean result = false;
         if (containsShortName(supportContextStrategyChecks, event)
-                || containsShortName(checkNamesForContextStrategyByTokenOrParentSet, event)) {
+                || containsShortName(checkNamesForContextStrategyByTokenOrAncestorSet, event)) {
             final DetailAST eventAst;
-            if (containsShortName(checkNamesForContextStrategyByTokenOrParentSet, event)) {
-                eventAst = getEventAst(event).getParent();
+            if (containsShortName(checkNamesForContextStrategyByTokenOrAncestorSet, event)) {
+                DetailAST tempAst = getEventAst(event).getParent();
+                while (!ancestorTokenSet.contains(tempAst.getType())) {
+                    tempAst = tempAst.getParent();
+                }
+                eventAst = tempAst;
             }
             else {
                 eventAst = getEventAst(event);
```

Comment:
please move this big if-else-block to a new method

Reply:
done